main.go

```diff
@@ -168,41 +169,57 @@ func restConfigForAPIExport(ctx context.Context, cfg *rest.Config, apiExportName
 		return nil, fmt.Errorf("error adding apis.kcp.dev/v1alpha1 to scheme: %w", err)
 	}
 
-	apiExportClient, err := client.New(cfg, client.Options{Scheme: scheme})
+	apiExportClient, err := client.NewWithWatch(cfg, client.Options{Scheme: scheme})
 	if err != nil {
 		return nil, fmt.Errorf("error creating APIExport client: %w", err)
 	}
 
-	var apiExport apisv1alpha1.APIExport
-
+	var opts []client.ListOption
 	if apiExportName != "" {
-		if err := apiExportClient.Get(ctx, types.NamespacedName{Name: apiExportName}, &apiExport); err != nil {
-			return nil, fmt.Errorf("error getting APIExport %q: %w", apiExportName, err)
-		}
-	} else {
-		setupLog.Info("api-export-name is empty - listing")
-		exports := &apisv1alpha1.APIExportList{}
-		if err := apiExportClient.List(ctx, exports); err != nil {
-			return nil, fmt.Errorf("error listing APIExports: %w", err)
-		}
-		if len(exports.Items) == 0 {
-			return nil, fmt.Errorf("no APIExport found")
-		}
-		if len(exports.Items) > 1 {
-			return nil, fmt.Errorf("more than one APIExport found")
-		}
-		apiExport = exports.Items[0]
+		opts = append(opts, client.MatchingFieldsSelector{
+			Selector: fields.OneTermEqualSelector("metadata.name", apiExportName),
+		})
 	}
 
-	if len(apiExport.Status.VirtualWorkspaces) < 1 {
-		return nil, fmt.Errorf("APIExport %q status.virtualWorkspaces is empty", apiExportName)
+	watch, err := apiExportClient.Watch(ctx, &apisv1alpha1.APIExportList{}, opts...)
+	if err != nil {
+		return nil, fmt.Errorf("error watching for APIExport: %w", err)
 	}
 
-	cfg = rest.CopyConfig(cfg)
-	// TODO(ncdc): sharding support
-	cfg.Host = apiExport.Status.VirtualWorkspaces[0].URL
+	for {
+		select {
+		case <-ctx.Done():
+			return nil, ctx.Err()
+		case e := <-watch.ResultChan():
+			apiExport, ok := e.Object.(*apisv1alpha1.APIExport)
+			if !ok {
+				continue
+			}
+
+			if !slices.Contains(apiExport.Spec.LatestResourceSchemas, "today.widgets.data.my.domain") {
+				// This is not this controller APIExport
```

Comment:
I think this check is likely superfluous and if it were to be true, the controller can do nothing except for panic and process exit, right?

Reply:
> We need to do a GET or LIST first to see the current state (and short-circuit if the live read shows a URL), and use that resource version to start the watch. Also, likely best to have watch resumption on errors and re-list on compaction - which makes me wonder if we should instead start an informer.

Ah, I have assumed this is what the controller runtime cached client actually does by default for typed objects. Let me check.

Reply:
This may be useful for the case where the APIExport name is not provided, as a way to pick the expected one. That's true the controller won't be able to to anything if the APIExport specified by name does not pass that check. So maybe add that check only when the name is not provided?

Reply:
If no name is specified, won't the watch with the `metadata.name` field selector return nothing?

Reply:
The `metadata.name` field selector is only appended when the name is provided. We could skip the check as well in that case or fail-fast.

Reply:
I've added the `apiExportName == ""` condition to the check.

Reply:
We should just make it required - why not?

Reply:
I took it as an invariant that was already there, but making it required, possibly with a sensible default, would make sense and simplify things.

Reply:
I've updated the `api-export-name` option to default to the controller APIExport name, so it avoids watching for multiple resources.